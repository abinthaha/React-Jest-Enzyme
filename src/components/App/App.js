import React, { Component } from 'react';
import { Button } from 'react-bootstrap';
import GiftComponent from '../Gifts/Gifts';

import { max_number, array_check } from '../../helper';

class App extends Component {

    constructor(props) {
        super(props);
        this.state = {
            gifts: []
        }
    }

    addGifts = () => {
        const { gifts } = this.state;
        gifts.push({id: max_number(this.state.gifts.map(gift => gift.id)) + 1})
        this.setState({
            ...this.state,
            gifts
        })
    }

    removeGift = id => {
        const filteredList = this.state.gifts.filter(gift => {
            return gift.id !== id
        })
        this.setState({
            ...this.state,
            gifts: array_check(filteredList)
        })
    }

    render() {
        return (
            <section>
                <h2>Gift Giver</h2>
                <div className='gift-list'>
                    {
                        this.state.gifts.map (gift => {
                            return (
                                <GiftComponent 
                                    key={gift.id}
                                    gift={gift}
                                    removeGift={this.removeGift}
                                />
                            )
                        })
                    }
                </div>
                <Button
                    className='btn-add'
                    onClick={ev => this.addGifts()}
                >
                    Add Gift
                </Button>
            </section>
        )
    }
}

export default App;