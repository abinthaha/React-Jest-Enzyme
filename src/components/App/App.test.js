import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

describe('App', () => {
    const app = shallow( <App /> );

    it('Renders correctly', () => {
        expect(app).toMatchSnapshot();
    })

    it('Initializes the `state` with and empty list of gifts', () => {
        expect(app.state().gifts).toEqual([])
    })

    describe('When clicking the `add-gift` button', () => {

        const id = 1;
        
        beforeEach(() => {
            app.find('.btn-add').simulate('click');
        });

        afterEach(() => {
            app.setState({
                gifts: []
            });
        })

        it('Add a new gift to the gifts `state` when clicking the `add gift` button', () => {
            expect(app.state().gifts).toEqual([{ id: id }]);
        });
    
        it('Adds a new gift to the gifts on clicking the `add gift` button', () => {
            expect(app.find('.gift-list').children().length).toEqual(1)
        });

        it('creates a gift', () => {
            expect(app.find('GiftComponent').exists()).toBe(true);
        })

        describe('User wants to remove a  gift from gifts', () => {
            beforeEach(() => {
                app.instance().removeGift(id);
            })

            it('Removes the gift from the `state`', () => {
                expect(app.state().gifts).toEqual([])
            })
        })
    })
})