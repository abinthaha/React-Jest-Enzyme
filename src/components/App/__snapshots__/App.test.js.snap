// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Renders correctly 1`] = `
<section>
  <h2>
    Gift Giver
  </h2>
  <div
    className="gift-list"
  />
  <ForwardRef(Bootstrap(Button))
    className="btn-add"
    onClick={[Function]}
  >
    Add Gift
  </ForwardRef(Bootstrap(Button))>
</section>
`;
